geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/stats/GeoMesaStats.scala

```diff
@@ -0,0 +1,167 @@
+        .getOrElse(wholeWorldEnvelope)
+
+  /**
+   * Get rough time bounds for a query
+   * Note: we don't currently filter by the cql
+   *
+   * @param query query
+   * @return time bounds
+   */
+  override def estimateTimeBounds(query: Query): Interval =
+    readTimeBounds(query.getTypeName).getOrElse(allTimeBounds)
+
+  /**
+   * Rough estimate of number of features
+   * Note: we don't currently filter by the cql
+   *
+   * @param query query
+   * @return count of features
+   */
+  override def estimateCount(query: Query): Long =
+    metadata.getTableSize(getTableName(query.getTypeName, RecordTable))
+
+  /**
+   * Clears any existing spatial bounds
+   *
+   * @param typeName simple feature type
+   */
+  def clearSpatialBounds(typeName: String): Unit = metadata.insert(typeName, SPATIAL_BOUNDS_KEY, "")
+
+  /**
+   * Clears any existing temporal bounds
+   *
+   * @param typeName simple feature type
+   */
+  def clearTemporalBounds(typeName: String): Unit = metadata.insert(typeName, TEMPORAL_BOUNDS_KEY, "")
+
+  /**
+   * Writes spatial bounds for this feature
+   *
+   * @param typeName simple feature type
+   * @param bounds partial bounds - existing bounds will be expanded to include this
+   */
+  def writeSpatialBounds(typeName: String, bounds: Envelope): Unit = {
+    val toWrite = readSpatialBounds(typeName) match {
+      case None => Some(bounds)
+      case Some(current) =>
+        val expanded = new Envelope(current)
+        expanded.expandToInclude(bounds)
+        if (current == expanded) None else Some(expanded)
+    }
+    toWrite.foreach(b => metadata.insert(typeName, SPATIAL_BOUNDS_KEY, encode(b)))
+  }
+
+  /**
+   * Writes temporal bounds for this feature
+   *
+   * @param typeName simple feature type
+   * @param bounds partial bounds - existing bounds will be expanded to include this
+   */
+  def writeTemporalBounds(typeName: String, bounds: Interval): Unit = {
+    import org.locationtech.geomesa.utils.time.Time.RichInterval
+    val toWrite = readTimeBounds(typeName) match {
+      case None => Some(bounds)
+      case Some(current) =>
+        val expanded = current.expandByInterval(bounds)
+        if (current == expanded) None else Some(expanded)
+    }
+    toWrite.foreach(b => metadata.insert(typeName, TEMPORAL_BOUNDS_KEY, encode(b)))
+  }
+
+  private def readTimeBounds(typeName: String): Option[Interval] =
+    metadata.readNoCache(typeName, TEMPORAL_BOUNDS_KEY).filterNot(_.isEmpty).map(decodeTimeBounds)
+
+  private def readSpatialBounds(typeName: String): Option[Envelope] =
+    metadata.readNoCache(typeName, SPATIAL_BOUNDS_KEY).filterNot(_.isEmpty).map(decodeSpatialBounds)
+}
+
+object GeoMesaStats {
+
+  def allTimeBounds = new Interval(0L, System.currentTimeMillis(), DateTimeZone.UTC) // Epoch till now
+
+  def decodeTimeBounds(value: String): Interval = {
+    val longs = value.split(":").map(java.lang.Long.parseLong)
+    require(longs(0) <= longs(1))
+    require(longs.length == 2)
```

Comment:
prob move the len == 2 check before the index lookup...als maybe add short error message